core/base/src/TSystem.cxx

```diff
@@ -3017,18 +3029,23 @@ int TSystem::CompileMacro(const char *filename, Option_t *opt,
    library = gSystem->UnixPathName(library);
 
    // ======= Check if the library need to loaded or compiled
-   if (!gInterpreter->IsLibraryLoaded(library) && gInterpreter->IsLoaded(expFileName)) {
-      // the script has already been loaded in interpreted mode
-      // Let's warn the user and unload it.
+   {
+      // The next block may lock the interpreter up to three times. Lock it once
+      // here instead.
+      R__LOCKGUARD(gGlobalMutex); // NEWLY_ADDED
```

Comment:
The routine cover here (`TCling's IsLibraryLoaded, IsLoaded and UnloadedFile`) should be made thread safe on their own.  (Other users might use them directly)